### PR TITLE
Specialize Avg and Sum accumulators (#6842)

### PR DESCRIPTION
**Pull request by tustvold**
## Which issue does this PR close?

Part of #6842

## Rationale for this change

This makes it easier to see what is going on, and avoids using ScalarValue arithmetic

### CODE REVIEW

File: datafusion/physical-expr/src/aggregate/average.rs [hunk 1]
```diff
@@ -208,97 +217,164 @@ impl PartialEq<dyn Any> for Avg {
...
 
     fn retract_batch(&mut self, values: &[ArrayRef]) -> Result<()> {
-        let values = &values[0];
+        let values = self.cast_input(&values[0])?;
+        let values = values.as_primitive::<Float64Type>();
         self.count -= (values.len() - values.null_count()) as u64;
-        let delta = sum_batch(values, &self.sum.get_datatype())?;
-        self.sum = self.sum.sub(&delta)?;
+        if let Some(x) = sum(values) {
+            self.sum = Some(self.sum.unwrap() - x);
```
Comment (tustvold) [inline datafusion/physical-expr/src/aggregate/average.rs:271]: This feels instinctively wrong, as it will accumulate errors over time... I'm not really sure what to do about this though...

File: datafusion/physical-expr/src/aggregate/average.rs [hunk 2]
```diff
@@ -493,104 +569,111 @@ mod tests {
...
     use crate::expressions::col;
     use crate::expressions::tests::aggregate;
-    use crate::generic_test_op;
     use arrow::record_batch::RecordBatch;
     use arrow::{array::*, datatypes::*};
-    use datafusion_common::Result;
+    use datafusion_expr::aggregate_function::sum_type_of_avg;
+    use datafusion_expr::type_coercion::aggregates::avg_return_type;
+
+    fn test_with_pre_cast(array: ArrayRef, expected: ScalarValue) {
```
Comment (tustvold) [inline datafusion/physical-expr/src/aggregate/average.rs:577]: This change is necessary because generic_test_op would call Avg::new which would then not have the correct arguments

File: datafusion/physical-expr/src/aggregate/average.rs [hunk 3]
```diff
@@ -106,11 +100,29 @@ impl AggregateExpr for Avg {
...
-        Ok(Box::new(AvgAccumulator::try_new(
-            // avg is f64 or decimal
-            &self.sum_data_type,
-            &self.rt_data_type,
-        )?))
+        use DataType::*;
+        // instantiate specialized accumulator based for the type
+        match (&self.sum_data_type, &self.rt_data_type) {
+            (Float64, Float64) => {
+                Ok(Box::new(AvgAccumulator::new(self.pre_cast_to_sum_type)))
```
Comment (tustvold) [inline datafusion/physical-expr/src/aggregate/average.rs:107]: This whole precast thing seems like a hack, imo this should be being handled by the type coercion machinery, not internal to the aggregator...

File: datafusion/physical-expr/src/aggregate/average.rs (hunk 1 above)
Comment (Dandandan) [reply datafusion/physical-expr/src/aggregate/average.rs:271]: I think this is expected for floats (otherwise we would need to keep intermediate values). Switching to decimal should allow for precise values.

FYI @ozankabak @metesynnada

Comment (metesynnada) [reply datafusion/physical-expr/src/aggregate/average.rs:271]: We will be looking into this.

**Comment by tustvold** [comment]
Marking as draft as I'd like to get #7369 in first

File: datafusion/physical-expr/src/aggregate/average.rs (hunk 1 above)
Comment (ozankabak) [reply datafusion/physical-expr/src/aggregate/average.rs:233]: The calculation looks correct. There is indeed some error accumulation when doing incremental calculations, but it is unavoidable (and very very rarely causes issues in practice)

File: datafusion/physical-expr/src/aggregate/sum_distinct.rs [hunk 4]
```diff
@@ -106,29 +116,56 @@ impl PartialEq<dyn Any> for DistinctSum {
...
-#[derive(Debug)]
-struct DistinctSumAccumulator {
-    hash_values: HashSet<ScalarValue, RandomState>,
+/// A wrapper around a type to provide hash for floats
+#[derive(Copy, Clone)]
+struct Hashable<T>(T);
+
+impl<T: ToByteSlice> std::hash::Hash for Hashable<T> {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        self.0.to_byte_slice().hash(state)
```
Comment (Dandandan) [inline datafusion/physical-expr/src/aggregate/sum_distinct.rs:125]: In other cases I think we only do this for floats and use the `state.hash_one(self.0)` in other cases (also preferably use `ahash` (it is faster for primitives).

Comment (tustvold) [reply datafusion/physical-expr/src/aggregate/sum_distinct.rs:125]: This does use ahash, it overrides the BuildHasher used by the HashSet. I think we could definitely do something better here, but in the absence of benchmarks I'm keen to go with simple and we can always optimise it later down the line. Regardless this should be significantly faster than the prior approach

Comment (Dandandan) [reply datafusion/physical-expr/src/aggregate/sum_distinct.rs:125]: > Edit: If someone really care about the performance of DistinctSum, implementing a GroupsAccumulator will likely yield far greater performance than any incremental tweaking of this Accumulator-based version

yeah true :)

File: datafusion/physical-expr/src/aggregate/sum_distinct.rs [hunk 5]
```diff
@@ -84,7 +89,12 @@ impl AggregateExpr for DistinctSum {
     }
 
     fn create_accumulator(&self) -> Result<Box<dyn Accumulator>> {
-        Ok(Box::new(DistinctSumAccumulator::try_new(&self.data_type)?))
+        macro_rules! helper {
+            ($t:ty, $dt:expr) => {
+                Ok(Box::new(DistinctSumAccumulator::<$t>::try_new(&$dt)?))
```
Comment (Dandandan) [inline datafusion/physical-expr/src/aggregate/sum_distinct.rs:94]: I think we can do the same for `DistinctCountAccumulator`

Comment (tustvold) [reply datafusion/physical-expr/src/aggregate/sum_distinct.rs:94]: Quite possibly, right now I'm just doing the minimum to be able to remove the ScalarValue arithmetic kernels :sweat_smile:

File: datafusion/physical-expr/src/aggregate/average.rs [hunk 6]
```diff
@@ -93,11 +87,27 @@ impl AggregateExpr for Avg {
...
+            ) => Ok(Box::new(DecimalAvgAccumulator {
+                sum: None,
+                count: 0,
+                sum_scale: *sum_scale,
+                sum_precision: *sum_precision,
+                target_precision: *target_precision,
+                target_scale: *target_scale,
+            })),
+            _ => not_impl_err!(
+                "AvgGroupsAccumulator for ({} --> {})",
```
Comment (alamb) [inline datafusion/physical-expr/src/aggregate/average.rs:106]:
```suggestion
                "AvgAccumulator for ({} --> {})",
```

File: datafusion/physical-expr/src/aggregate/sum.rs [hunk 7]
```diff
@@ -123,54 +122,28 @@ impl AggregateExpr for Sum {
...
-            ),
+        macro_rules! helper {
+            ($t:ty, $dt:expr) => {
+                Ok(Box::new(PrimitiveGroupsAccumulator::<$t, _>::new(
+                    &$dt,
+                    |x, y| *x = x.add_wrapping(y),
+                )))
+            };
         }
+        downcast_sum!(self, helper)
```
Comment (alamb) [inline datafusion/physical-expr/src/aggregate/sum.rs:133]: the multiple levels of macros is concise I'll give you that but I do find it hard to follow. Maybe that is ok as we don't expect this to be changing

File: datafusion/physical-expr/src/aggregate/sum.rs [hunk 8]
```diff
@@ -71,18 +63,20 @@ impl Sum {
...
-///
-/// [`ArrowPrimitiveType`]: arrow::datatypes::ArrowPrimitiveType
-macro_rules! instantiate_primitive_accumulator {
-    ($SELF:expr, $PRIMTYPE:ident, $FN:expr) => {{
-        Ok(Box::new(PrimitiveGroupsAccumulator::<$PRIMTYPE, _>::new(
-            &$SELF.data_type,
-            $FN,
-        )))
-    }};
+/// Sum only supports a subset of numeric types, instead relying on type coercion
```
Comment (alamb) [inline datafusion/physical-expr/src/aggregate/sum.rs:66]: It might help to document what this macro does (aka calls helper macro given a s (what is s? The aggregate?))

**Review by alamb — APPROVED** [review]
Thank you @tustvold  -- these changes make sense to me.

 I think we should run some basic performance tests too if we have any that cover queries that use these functions (like SUM DISTINCT and using sliding sum in window functions). Maybe @metesynnada or @ozankabak know of some benchmarks we can run that cover it

**Comment by ozankabak** [comment]
> I think we should run some basic performance tests too if we have any that cover queries that use these functions (like SUM DISTINCT and using sliding sum in window functions). Maybe @metesynnada or @ozankabak know of some benchmarks we can run that cover it

We will discuss this tomorrow and circle back to you

File: datafusion/physical-expr/src/aggregate/sum_distinct.rs (hunk 5 above)
Comment (tustvold) [reply datafusion/physical-expr/src/aggregate/sum_distinct.rs:94]: But not arithmetic :smile:

Comment (Dandandan) [reply datafusion/physical-expr/src/aggregate/sum_distinct.rs:94]: Aha 😁 let me follow up this PR then ;)

**Review by Dandandan — APPROVED** [review]
Looks great 👍

**Comment by tustvold** [comment]
Running the TPCH benchmarks

```
┏━━━━━━━━━━━━━━┳━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Query        ┃      main ┃ specialize-avg-accumulator ┃        Change ┃
┡━━━━━━━━━━━━━━╇━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ QQuery 1     │  671.28ms │                   653.98ms │     no change │
│ QQuery 2     │  147.40ms │                   136.89ms │ +1.08x faster │
│ QQuery 3     │  259.99ms │                   252.70ms │     no change │
│ QQuery 4     │  150.08ms │                   145.15ms │     no change │
│ QQuery 5     │  347.39ms │                   348.76ms │     no change │
│ QQuery 6     │  152.53ms │                   147.19ms │     no change │
│ QQuery 7     │  595.27ms │                   560.30ms │ +1.06x faster │
│ QQuery 8     │  382.90ms │                   368.43ms │     no change │
│ QQuery 9     │  624.58ms │                   593.67ms │     no change │
│ QQuery 10    │  462.84ms │                   452.43ms │     no change │
│ QQuery 11    │  138.77ms │                   127.82ms │ +1.09x faster │
│ QQuery 12    │  221.17ms │                   213.70ms │     no change │
│ QQuery 13    │  389.77ms │                   375.43ms │     no change │
│ QQuery 14    │  209.97ms │                   209.11ms │     no change │
│ QQuery 15    │  149.18ms │                   144.03ms │     no change │
│ QQuery 16    │  134.44ms │                   114.72ms │ +1.17x faster │
│ QQuery 17    │  707.01ms │                   694.61ms │     no change │
│ QQuery 18    │ 1066.54ms │                  1086.68ms │     no change │
│ QQuery 19    │  384.12ms │                   367.69ms │     no change │
│ QQuery 20    │  343.48ms │                   317.19ms │ +1.08x faster │
│ QQuery 21    │  850.10ms │                   833.78ms │     no change │
│ QQuery 22    │  107.96ms │                   100.84ms │ +1.07x faster │
└──────────────┴───────────┴────────────────────────────┴───────────────┘
```
